### PR TITLE
Add batching support for evaluation

### PR DESCRIPTION
**Pull request by athitten**
# What does this PR do ?

1) Adds batching support in `loglikelihood` method of `nemo/collections/llm/evaluation/base.py` and query's the model with batched inputs. Introduces `batch_size` arg in the [evaluate](https://github.com/NVIDIA/NeMo/pull/11934/files#diff-1afe0432f8c676bf15b5ebbc6628d0f1e9757ea6d3eb73a458c905080e77d50bR438) (in `nemo/collections/llm/api.py`) method for batching. Also reintroduces `max_tokens_to_generate` as its required for `generate_until` type tasks (ex: gsm8k)
2) In `nemo/export/tensorrt_llm.py` file, adds a private method `_pad_logits` that performs padding for context logits when they are available, since in case of batched inputs the returned context_logits can have varying seq_len and we need padding to convert to numpy array.
3) Other minor edits for which relevant comments have been added.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/evaluation/base.py [hunk 1]
```diff
@@ -54,12 +57,13 @@ def _generate_tokens_logits(
...
-        else:
-            # In case of multiple token prediction return the context logits
-            output_context_logits = True
+        if return_logits:  # in case of loglikelihood type tasks
+            if single_prediction_token:
+                # In case of single token prediction like mmlu return only the generation logits
+                output_generation_logits = True
+            else:
+                # In case of multiple token prediction return the full context logits
+                output_context_logits = True
```
Comment (athitten) [inline nemo/collections/llm/evaluation/base.py:66]: `output_generation_logits` and `output_context_logits` are false by default for generate_until type task as it needs only text and not logits. Modify them only if we are returning `logits`.

File: nemo/collections/llm/evaluation/base.py [hunk 2]
```diff
@@ -107,62 +111,91 @@ def loglikelihood(self, requests: list[Instance]):
...
-        # task.
+        # Assuming evaluating on only one benchmark/task at a time, hence all instances in requests are of the same task.
         mmlu_regex_pattern = r"^mmlu_"
-        lambada_regex_pattern = r"^lambada_"
-        if re.match(mmlu_regex_pattern, requests[0].task_name) or re.match(
-            lambada_regex_pattern, requests[0].task_name
-        ):
+        if re.match(mmlu_regex_pattern, requests[0].task_name):
+            # in case of mmlu the output token is one of 'a','b','c','d'
             single_prediction_token = True
```
Comment (athitten) [inline nemo/collections/llm/evaluation/base.py:119]: Categorize only mmlu as single token prediction task since in case of lambada, although its a single token prediction task, the tokenized output can be multi tokens depending on how the tokenizer splits it. MMLU is multiple choice and output is always one of 'a','b','c','d'.

File: nemo/collections/llm/evaluation/base.py [hunk 3]
```diff
@@ -107,62 +111,91 @@ def loglikelihood(self, requests: list[Instance]):
...
-            continuation_enc = self.tokenizer.tokenizer.encode(continuation, **special_tokens_kwargs)
-            # for SentencePeice consider the encoded tokens from the 2nd token since first encoded token is space.
-            if self.tokenizer_type(self.tokenizer) == "SentencePieceTokenizer":
-                continuation_enc = continuation_enc[1:]
-            num_cont_tokens = len(continuation_enc)
-            # Hard code max_tokens_to_generate to 1 to always generate just 1 token
-            self.max_tokens_to_generate = 1
-            # Delete the last token from continuation before passing it to the ip prompt by replacing with empty string
-            prompt = context + continuation.replace(self.tokenizer.tokenizer.decode(continuation_enc[-1]), "")
-            # Create payload to query the model deployed on PyTriton server
```
Comment (athitten) [inline nemo/collections/llm/evaluation/base.py:135]: batching logic below.

File: nemo/collections/llm/evaluation/base.py [hunk 4]
```diff
@@ -33,19 +34,21 @@ class NeMoFWLMEval(LM):
...
         self.temperature = temperature
         self.top_p = top_p
         self.top_k = top_k
         self.add_bos = add_bos
         super().__init__()
 
-    def _generate_tokens_logits(
-        self, payload, single_prediction_token, return_text: bool = False, return_logits: bool = False
-    ):
+    def _generate_tokens_logits(self, payload, single_prediction_token: bool = False, return_logits: bool = False):
```
Comment (athitten) [inline nemo/collections/llm/evaluation/base.py:51]: Remove `return_text` arg, as its redundant and can be handled with just `return_logits` (False by default) and made True for loglikelihood tasks and remains false for `generate_until` type task.

File: nemo/collections/llm/api.py [hunk 5]
```diff
@@ -387,11 +387,6 @@ def deploy(
 
     unset_environment_variables()
 
-    if not isinstance(nemo_checkpoint, Path):
-        nemo_checkpoint = Path(nemo_checkpoint)
-    if not isinstance(triton_model_repository, Path):
-        triton_model_repository = Path(triton_model_repository)
-
```
Comment (athitten) [inline nemo/collections/llm/api.py:394]: Not required as export to trtllm expects the path to be str and doesn't need to be a Path object. Same for `triton_model_repository` path.

**Review by oyilmaz-nvidia — APPROVED** [review]

**Comment by Glorf** [comment]
@athitten overall looks good! Did you manage to run any regression testing between the two versions? Would be great to see if the results stay the same :)

**Review by Glorf — APPROVED** [review]

**Comment by athitten** [comment]
> @athitten overall looks good! Did you manage to run any regression testing between the two versions? Would be great to see if the results stay the same :)

Yes please the see the accuracy results below comparing the code in NeMo with batching and running evaluations directly with lm-eval-harness on llama3-8b model
Screen Shot 2025-01-28 at 2 29 21 PM
